tools/envoy_headersplit/headersplit.py

```diff
@@ -0,0 +1,378 @@
+  for i, cursor in enumerate(impl_cursors):
+    classname = cursor.semantic_parent.spelling
+    # get first line of function body
+    implline = get_implline(cursor)
+    # get last line of function body
+    if i + 1 < len(impl_cursors):
+      # i is not the last method, get the start line for the next method
+      # as the last line of i
+      _, impl_end = get_implline(impl_cursors[i + 1])
+      impl = "".join(source_code[implline:impl_end])
+    else:
+      # i is the last method, after removing the lines containing close brackets
+      # for namespaces, the rest should be the function body
+      offset = 0
+      while implline + offset < len(source_code):
+        if "// namespace" in source_code[implline + offset]:
+          break
+        offset += 1
+      impl = "".join(source_code[implline:implline + offset])
+    if classname in classname_to_impl:
+      classname_to_impl[classname] += impl + "\n"
+    else:
+      classname_to_impl[classname] = impl + "\n"
+  return classname_to_impl
+
+
+def get_enclosing_namespace(defn: Cursor) -> Tuple[str, str]:
+  """
+    retrieves all enclosing namespaces for the class pointed by defn.
+
+    this is necessary to construct the mock class header
+
+    e.g.
+
+    defn is pointing MockClass in the follow source code:
+
+    namespace Envoy {
+    namespace Server {
+    class MockClass2 {...}
+    namespace Configuration {
+    class MockClass {...}
+          ^ 
+          defn
+    }
+    }
+    }
+
+    this function will return "namespace Envoy {\nnamespace Server {\nnamespace Configuration{\n" and "\n}\n}\n}\n" 
+
+    Args:
+        defn: libclang Cursor pointing to a mock class
+
+    Returns:
+        namespace_prefix, namespace_suffix: a pair of string, representing the enclosing namespaces
+    """
+  namespace_prefix = ""
+  namespace_suffix = ""
+  parent_cursor = defn.semantic_parent
+  while parent_cursor.kind == CursorKind.NAMESPACE:
+    if parent_cursor.spelling == "":
+      break
+    namespace_prefix = "namespace {} {{\n".format(parent_cursor.spelling) + namespace_prefix
+    namespace_suffix += "\n}"
+    parent_cursor = parent_cursor.semantic_parent
+  namespace_suffix += "\n"
+  return namespace_prefix, namespace_suffix
+
+
+def read_file_contents(path):
+  with open(path, "r") as input_file:
+    return input_file.read()
+
+
+def write_file_contents(class_name, class_defn, class_impl):
+  with open("{}.h".format(to_filename(class_name)), "w") as decl_file:
+    decl_file.write(class_defn)
+  with open("{}.cc".format(to_filename(class_name)), "w") as impl_file:
+    impl_file.write(class_impl)
+  # generating bazel build file, need to fill dependency manually
+  bazel_text = """
+envoy_cc_mock(
+  name = "{}_mocks",
+  srcs = ["{}.cc"],
+  hdrs = ["{}.h"],
+  deps = [
+
+  ]
+)
+""".format(to_filename(class_name), to_filename(class_name), to_filename(class_name))
+  with open("BUILD", "a") as bazel_file:
+    bazel_file.write(bazel_text)
```

Comment:
If this is append, this script is somewhat non-idempotent. What happens if you run it multiple times?

Reply:
it will add rule for same mock class target for multiple times.
Now I add statements to check if the target has been already appended. So now it will only add once if run multiple times